### DISCUSSION TITLE
Saving UploadFile object to S3 with upload_fileobj

### DISCUSSION THREAD

**Discussion by jcf-dev** [fastapi/fastapi#6906] [Questions] [answered]
I've been scratching my head on this for days now. I still couldn't solve the problem. Basically, I just wanted to put a CSV file in a LocalStack S3, and I can't get it working.

Here's the snippet of my code:

api.py

        from files import s3, AWS_S3_BUCKET_NAME, upload_file_to_bucket
            
        @router.post('/api/customer/generate/upload',
                         name='Upload CSV to AWS S3 Bucket',
                         status_code=201)
            async def post_upload_user_csv(file_obj: UploadFile = File(...)):
                upload_obj = upload_file_to_bucket(s3_client=s3(),
                                                   file_obj=file_obj.file,
                                                   bucket=AWS_S3_BUCKET_NAME,
                                                   folder='CSV',  # To Be updated
                                                   object_name=file_obj.filename)
                if upload_obj:
                    return JSONResponse(content="Object has been uploaded to bucket successfully",
                                        status_code=status.HTTP_201_CREATED)
                else:
                    raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                                        detail="File could not be uploaded")

files.py

      import os
      import boto3
      import logging
      
      from botocore.client import BaseClient
      from botocore.exceptions import ClientError
      
      AWS_ACCESS_KEY_ID = os.getenv('POSTGRES_HOST')
      AWS_SECRET_KEY = os.getenv('AWS_SECRET_KEY')
      AWS_S3_BUCKET_NAME = os.getenv('AWS_S3_BUCKET_NAME')
      
      
      def s3() -> BaseClient:
          client = boto3.client(service_name='s3',
                                aws_access_key_id=AWS_ACCESS_KEY_ID,
                                aws_secret_access_key=AWS_SECRET_KEY,
                                endpoint_url='http://localhost:4566/')  # Use LocalStack Endpoint
      
          return client
      
      
      def upload_file_to_bucket(s3_client, file_obj, bucket, folder, object_name=None):
          """Upload a file to an S3 bucket
      
          :param s3_client: S3 Client
          :param file_obj: File to upload
          :param bucket: Bucket to upload to
          :param folder: Folder to upload to
          :param object_name: S3 object name. If not specified then file_name is used
          :return: True if file was uploaded, else False
          """
          # If S3 object_name was not specified, use file_name
          if object_name is None:
              object_name = file_obj
      
          # Upload the file
          try:
              # with open("files", "rb") as f:
              s3_client.upload_fileobj(file_obj, bucket, f"{folder}/{object_name}")
          except ClientError as e:
              logging.error(e)
              return False
          return True

The problem is that s3_client needs to open the file in binary mode first before I can upload it to the s3 bucket. However, this can't be done directly and the file needs to be saved temporarily on the FastAPI server but I really don't want to do that for obvious reasons.

Any help will be much appreciated. Thank you in advance!

**Comment by jcf-dev** [comment]
Sorry for the confusion at the last, I just realized it now. 

Please correct me I got something wrong with my logic. Of course, writing the file to a variable is very much preferred rather than saving it on a disk. 

From what I understand, UploadFile is a file-like object, meaning it's stored temporarily on the memory and not directly on the disk. What I'm trying to do is chunk by chunk or asynchronously upload the provided file directly to S3, whenever possible

**Comment by jgontrum** [comment, marked as answer]
I just came across your issue and maybe I can help you a bit, though I haven't figured out a full solution for this myself.

1. The way I understand UploadFile, it is storing the uploaded file in memory first, but uses a "SpooledTemporaryFile" internally that writes the content on disk if the upload gets too large.

2. One way to circumvent this is by reading chunks of the uploaded file and sending them to S3 chunk by chunk. This could be done like that:

```python
@router.post("/upload")
async def upload(file: UploadFile = File(...)):
    upload_file = io.BytesIO()
    while True:
        chunk = await file.read(1024)
        if not chunk:
            break
        upload_file.write(chunk)

```

The problem I'm facing now is that S3 doesn't seem to support streaming the upload. But maybe this is a step in the right direction?

**Comment by jcf-dev** [comment]
> I just came across your issue and maybe I can help you a bit, though I haven't figured out a full solution for this myself.
> 
> 1. The way I understand UploadFile, it is storing the uploaded file in memory first, but uses a "SpooledTemporaryFile" internally that writes the content on disk if the upload gets too large.
> 2. One way to circumvent this is by reading chunks of the uploaded file and sending them to S3 chunk by chunk. This could be done like that:
> 
> ```python
> @router.post("/upload")
> async def upload(file: UploadFile = File(...)):
>     upload_file = io.BytesIO()
>     while True:
>         chunk = await file.read(1024)
>         if not chunk:
>             break
>         upload_file.write(chunk)
> ```
> 
> The problem I'm facing now is that S3 doesn't seem to support streaming the upload. But maybe this is a step in the right direction?

Hi. Thank you for your response, and yes it did put me in the right direction. 😉

**Comment by tiangolo** [comment]
Thanks for the help here @jgontrum ! 👏  🙇 

Thanks for reporting back and closing the issue @jcf-dev  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
